### PR TITLE
#163383177 Reset password via email

### PR DESCRIPTION
**Pull request by pbkabali**
#### What does this PR do?

Adds password reset functionality to the app where a user who wishes to reset their password can request a password reset link which is sent directly into their registered mailbox. Users do not have to be signed in making this a great feature for forgotten passwords.

#### Description of tasks to be completed

- Create password reset link incorporating a token that is unique to the requesting user
- Send the password reset link to the requesting user's registered mailbox
- Allow the user to set a new password for their account when they follow the link in the email

#### How should this be manually tested?

Clone the repo into a virtualenv, cd into it and run:
`pip3 install -r requirements.txt`
`python3 manage.py makemigrations`
`python3 manage.py migrate`
`python3 manage.py runserver`

Using an API testing tool such as Postman, perform the following HTTP requests:
- Register a new user with a valid email address:
`ENDPOINT` `POST` `http://localhost:8000/api/v1/users/`
`BODY`  
`{"user":{"email":"your_email",`
`"username": "your_username",`
`"password":"your_password"}}`
`content-type: application/json`

- Try logging in to confirm your details are saved
`ENDPOINT` `POST` `http://localhost:8000/api/v1/users/login/`
`BODY`  
`{"user":{"email":"your_email",`
`"password":"your_password"}}`
`content-type: application/json`

- Request password reset link using your registered email address
`ENDPOINT` `POST` `http://localhost:8000/api/v1/users/request-password-reset/`
`BODY`  
`{"user":{"email":"your_email"}}`
`content-type: application/json`

- check in your mailbox for a password reset link

- Submit your new password by making request to reset link
`ENDPOINT` `POST` `reset link from mailbox`
`BODY`  
`{"user":{"new_password":"your_new_password"}}`
`content-type: application/json`

- log into app with new password to confirm password reset.

#### Background information
This feature is invaluable in instances where our users might forget their login credentials.

#### Hosted app
[Visit this app online here](https://ah-backend-summer-stagin-pr-22.herokuapp.com/)

#### What are the relevant pivotal tracker stories?
[#163383177](https://www.pivotaltracker.com/story/show/163383177)

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -154,3 +154,24 @@ class TwitterAuthSerializer(serializers.Serializer):
...
+    new_password = serializers.CharField(
+        max_length=128, allow_blank=False, write_only=True, required=True,
+        min_length=6
+    )
+    def validate(self, data):
+        new_password = data.get('new_password', None)
+        if not re.compile(r'^[0-9a-zA-Z]*$').match(new_password):
+            raise ValidationError(
+                "Password must contain alphanumeric characters only")
+        return new_password
```
Comment (malep2007) [inline authors/apps/authentication/serializers.py]: New line @pbkabali

File: authors/apps/authentication/tests/test_update.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ class UpdateTest(BaseTest):
 
     def setUp(self):
         super().setUp()
-        self.url_update = reverse('authentication:retrieve_update')
+        self.url_update = reverse('authentication:retrieve-update')
```
Comment (malep2007) [inline authors/apps/authentication/tests/test_update.py:15]: Why the change in url?

**Review by malep2007 — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/test_update.py (hunk 2 above)
Comment (missvicki) [reply authors/apps/authentication/tests/test_update.py:15]: This was done to follow the naming convention in the URL namespaces. we are using the hyphen(-) instead of the underscore(_). Hope this makes sense

File: authors/apps/core/password_reset_manager.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    password-reset feature including preparing the email message,
+    reset link and verification of link and updating password
+    """
+
+    def __init__(self, request):
+        self.sender_email = "teamabasama@gmail.com"
+        self.subject = "You requested a password reset"
+        self.password_reset_url = request.build_absolute_uri('/api/v1/users/reset-password/')
+
+    def get_user_by_email(self, email):
```
Comment (missvicki) [inline authors/apps/core/password_reset_manager.py:20]: @pbkabali  Could you please add descriptive docstring messages below the methods as well

Comment (pbkabali) [reply authors/apps/core/password_reset_manager.py:20]: @missvicki Working on that. Thanks  a lot

File: authors/apps/authentication/tests/test_update.py (hunk 2 above)
Comment (pbkabali) [reply authors/apps/authentication/tests/test_update.py:15]: I realize this required a separate PT story and shall follow the correct procedure the next time but just like @missvicki explained, we have agreed to follow the same convention of using hyphens and not underscores in our url-names.

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (pbkabali) [reply authors/apps/authentication/serializers.py]: Thanks for the observation @malep2007 .

**Review by malep2007 — APPROVED** [review]
